### PR TITLE
field reader

### PR DESCRIPTION
**Pull request by burrbull**
Generics part 2 (the easiest part)

r? @therealprof 

изображение

### CODE REVIEW

File: src/generate/generic.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+pub fn render() -> Result<Vec<Tokens>> {
+    let mut code = vec![];
+    let mut generic_items = vec![];
+
+    generic_items.push(quote! {
+        use core::marker;
+
+        ///Converting enumerated values to bits
+        pub trait ToBits<N> {
+            ///Convertion method
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
            ///Conversion method
```

File: src/generate/generic.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            fn _bits(&self) -> N;
+        }
+    });
+
+    generic_items.push(quote! {
+        ///Value read from the register
+        pub struct FR<U, T> {
```
Comment (therealprof) [inline src/generate/generic.rs]: Why use two types here? The code below always generates the same for both and just for `PhantomType` it shouldn't matter anyway.

Comment (burrbull) [reply src/generate/generic.rs]: Second type is used for enumerated values if they present and it's also needed for differ one reader from another.
If enumerated values are absent I just ignore second type.

Comment (therealprof) [reply src/generate/generic.rs]: > it's also needed for differ one reader from another

Huh?

> If enumerated values are absent I just ignore second type.

I noticed which makes me wonder what the purpose is. Where can I see an example of such an enumerated type?

Comment (burrbull) [reply src/generate/generic.rs]: https://docs.rs/stm32f1/0.7.1/stm32f1/stm32f103/gpioa/crh/enum.CNF8R.html#method.is_push_pull

Aside from `bits` method reader have also `is_push_pull` and other methods generated from `<enumeratedValues>` list.

Comment (burrbull) [reply src/generate/generic.rs]: Same with writers but more complex.

**Comment by burrbull** [comment]
I've updated changelog.

**Comment by therealprof** [comment]
bors retry

**Comment by therealprof** [comment]
Meh, the change mostly seems to improve compile time or not affect it at all but for Freescale it drives it over the 50 minute limit.

**Comment by burrbull** [comment]
|                   |  time  |   old  |
|:-----------------:|:------:|:------:|
|   MK02F12810.svd  | 48.86s | 48.64s |
|     MK10D7.svd    | 1m 02s | 50.72s |
|     MK11D5.svd    | 1m 22s | 1m 02s |
|     MK12D5.svd    | 1m 22s | 1m 06s |
| MK20D7.svd        | 1m 04s | 53.49s |
| MK21D5.svd        | 1m 27s | 1m 06s |
| MK21F12.svd       | 2m 10s | 1m 37s |
| MK22F12.svd       | 2m 11s | 1m 39s |
| MK22F51212.svd    | 1m 26s | 1m 12s |
| MK24F25612.svd    | 1m 06s | 1m 00s |
| MK30D7.svd        | 1m 17s | 1m 05s |
| MK40D7.svd        | 1m 22s | 1m 05s |
| MK50D7.svd        | 1m 03s | 53.44s |
| MK51D7.svd        | 1m 19s | 1m 04s |
| MK52DZ10.svd      | 1m 19s | 1m 00s |
| MK60D10.svd       | 1m 25s | 1m 03s |
| MK63F12.svd       | 1m 28s | 1m 05s |
| MK66F18.svd       | 1m 59s | 1m 22s |
| MK82F25615.svd    | 1m 49s | 1m 20s |
| MKE15Z7.svd       | 1m 30s | 1m 06s |
| MKL28T7_CORE0.svd | 2m 23s | 1m 29s |
| MKL81Z7.svd       | 1m 43s | 1m 16s |
| MKV10Z1287.svd    | 40.41s | 35.58s |
| MKV30F12810.svd   | 26.53s | 25.61s |
| MKV31F51212.svd   | 40.15s | 34.63s |
| MKV43F15.svd      | 31.72s | 29.63s |
| MKV45F15.svd      | 43.26s | 39.53s |
| MKW20Z4.svd       | 42.13s | 38.26s |
| MKW22D5.svd       | 1m 14s | 57.98s |
| MKW31Z4.svd       | 1m 04s | 55.94s |
| MKE02Z4.svd       | 12.71s | 12.88s |
| MKE06Z4.svd       | 16.37s | 16.18s |
| MKL02Z4.svd       | 7.35s  | 7.71s  |
| MKL05Z4.svd       | 10.11s | 10.06s |
| MKL15Z4.svd       | 16.81s | 16.67s |
| MKL17Z644.svd     | 46.03s | 39.11s |
| MKL26Z4.svd       | 46.20s | 38.93s |
| MKL33Z4.svd       | 1m 03s | 50.68s |
| MKL36Z4.svd       | 1m 01s | 48.82s |
| MKM14ZA5.svd      | 29.70s | 27.88s |
| MKM34ZA5.svd      | 41.74s | 38.49s |
| SKEAZN642.svd     | 13.07s | 13.42s |

**Comment by therealprof** [comment]
bors retry

**Comment by burrbull** [comment]
Rebased.

**Pull request merged**